날씨정보(후) [STEP2] Shapiro, Coden

ictechgy opened the pull request:
@YebinKim 
안녕하십니까 비비 ㅎㅎ.. 머지 1시간만에 바로 또 PR을 날려봅니다..! 🤣

이번 스텝에서도 구현해야 할 것이 많지 않았어서 질문도 별로 없습니다...!
그래도 잘 부탁드립니다!

## 구현내용
1. 화면 Orientation 전환
  이는 프로젝트 설정 변경을 통해 구현하였습니다! (`plist`가 바뀌었습니다.)

2. 지역화
  * 우선 한국어와 영어를 사용하는 지역을 중심으로 지역화를 구현하였습니다!
  * 스토리보드를 통해 화면을 구현하고 있지 않기 때문에 지역화를 위한 별도의 스토리보드 파일 등은 만들지 않았구요. 지역화를 위한 `Localizable string` 파일과 권한 요청창 지역화를 위한 `InfoPlist string` 파일을 만들었습니다. 
  * 온도 단위에 대한 지역화는 로컬에서 수행하기보다 API에서 `units` 파라미터로 제공해준다는 점을 발견하여 해당 기능을 사용하는 방식으로 구현하였습니다. 

## 궁금한 점
1. API 요청 시 `CommonWeatherAPIParameter` 타입 내의 `LanguageType` 프로퍼티 사용 여부
  API를 요청할 때 `LanguageType`을 통해 지역화를 관리 및 구현해보려 했습니다. 하지만 `LanguageType`을 지정하여 파라미터로 보낸다고 해도 response의 내용들은 딱히 바뀌지 않더라구요! API에 의존하기보다 저희가 따로 지역화를 해주는 방향으로 작성을 했는데 적절했을까요? 비비는 어떻게 생각하시나요??

shapiro711 committed:
feat: 화면 전환에도 대응되도록 프로젝트 설정 변경

ictechgy committed:
feat: TableView HeaderView의 주소 표기 방법을 지역화

shapiro711 committed:
feat: 지역에 따른 온도 단위 변화 로직 구현

- MeasurementType preferrdLanguage에 따라 MeasurementType을 얻을 수 있도록 getMeasurementType(by:) -> MeasurementType 메서드 구현
- 보간법 사용시 옵셔널 타입에 대해 Optional 키워드가 따라 붙게 되므로 generateURLQueryItem(name:value:) -> URLQueryItem? 메서드의 value 파라미터 타입 변경
- 헤더뷰와 셀에서 기존에 사용하던 온도 단위 변화 로직 삭제

ictechgy committed:
feat: Localization을 위한 Localizable string 파일 추가

- Localizable english 번역 추가
- Localization을 위해 프로젝트 설정에서 Localization 언어 추가

shapiro711 committed:
feat: Localizable Korean 파일 번역 추가

ictechgy committed:
feat: UI에 들어가는 String들 Localizing 하는 로직 추가

- Localizable string들에서 누락된 세미콜론 추가
- String Extension을 통해 NSLocalizedString에 대한 메서드 구현

shapiro711 committed:
feat: 권한 요청 Alert localization 구현

ictechgy committed:
refactor: 사용자 위치 얻어오기에 실패한 경우 띄워지는 에러 메시지 출력 방식 변경

shapiro711 committed:
refactor: APIable 프로토콜의 extention으로 기본값 제공

- WeatehrAPI, ImageAPI에서 기본 값이 제공된 프로퍼티 제거

ictechgy committed:
refactor: 위치 변경 Alert에서 사용자가 입력한 위경도를 이용하는 로직 변경

- 가독성을 위해 사용자가 입력한 위경도를 이용하여 CLLocation 인스턴스를 생성하는 부분 분리

shapiro711 committed:
refactor: MainWeatherTableViewCell identifier 설정

- 가독성을 위해 프로퍼티로 identifier를 가지고 있도록 변경

YebinKim commented on WeatherForecast/WeatherForecast/Extension/String + Extension.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+extension String {
```
`String` 의 확장으로 localized() 를 작성하셨군요! 멋진 활용입니다 😄
`NSLocalizedString` 생성자의 `comment` 파라미터도 전달할 수 있게 작성해두면 더욱 확장성있는 메서드가 될 수도 있을 거 같아요~

shapiro711 replied:
저희는 comment를 무조건 "" 로 지정하고 있었네요 !! 비비가 말씀해주신 대로 수정해볼게요~

YebinKim commented on WeatherForecast/WeatherForecast.xcodeproj/project.pbxproj:
```diff
@@ -142,6 +150,8 @@
 			isa = PBXGroup;
 			children = (
+				3F7591C6271DD33A00BF3F2E /* InfoPlist.strings */,
+				C9D93094271DC89C00BDBCF7 /* Localizable.strings */,
```
프로젝트의 가장 상위에 strings 파일들이 위치해있네요! 👀
프로젝트의 설정과 로컬라이징에 대한 폴더를 만들어서 따로 파일을 관리해보면 어떨까요?

ictechgy replied:
앗 넵! 프로젝트 파일들이 한눈에 들어올 수 있도록 폴더로 묶어서 관리해볼게요~~

YebinKim commented on WeatherForecast/WeatherForecast/Model/API/WeatherAPIable.swift:
```diff
@@ -79,6 +80,21 @@ enum MeasurementType: String {
     }
+    
+    static func getMeasurementType(by preferredLanguage: String?) -> MeasurementType {
+        guard let preferredLanguage = preferredLanguage  else {
```
else 앞에 띄어쓰기가 두 개 있는 거 같아요 👀

shapiro711 replied:
헛... 수정하겠습니다

YebinKim commented on WeatherForecast/WeatherForecast/View/MainWeatherHeaderView.swift:
```diff
@@ -128,17 +128,9 @@ final class MainWeatherHeaderView: UIView {
-        }
+        highestTemperatureLabel.text = "최고".localized() + (weatherData.mainWeatherInfomation?.maximumTemperature?.description ?? "")
+        lowestTemperatureLabel.text = "최저".localized() + (weatherData.mainWeatherInfomation?.minimumTemperature?.description ?? "")
+        currentTamperatureLabel.text = weatherData.mainWeatherInfomation?.temperature?.description
```
개인적인 의견인데..! 지역에 따라 섭씨와 화씨 기호를 표시해주면 어떨까요??
숫자만 쓰여 있어서 이게 온도를 나타내는 레이블인지, 섭씨 기준인지 화씨 기준인지를 알기가 어려운 거 같아서요 🤔

<img width="320" alt="label" src="https://user-images.githubusercontent.com/38066039/138125344-efba1405-592b-48d6-ac7f-b158c57dbe60.PNG">

shapiro711 replied:
아 넵 단위가 없으니까 어색하네요...! 수정해볼게요!!

YebinKim commented on WeatherForecast/ko.lproj/Localizable.strings:
```diff
@@ -0,0 +1,19 @@
+"최저" = "최저 ";
+"위치 변경" = "위치 변경";
+
+//Change Location Alert
+"변경할 좌표를 입력해주세요" = "변경할 좌표를 입력해주세요";
+"위도" = "위도";
+"경도" = "경도";
+"변경" = "변경";
+"현재 위치로 재설정" = "현재 위치로 재설정";
+"취소" = "취소";
```
로컬라이징 키를 한글로 사용하고 있네요! 🤔

한글로 키를 사용하게 되면 한 눈에 내용을 이해하기 쉽다는 장점이 있겠지만..
키에 대한 값이 변경되었을 경우 혼란이 생길 수 있기 때문에 저는 키값을 **해당 레이블의 역할에 대한 영문명**으로 정하고 있습니다!
예를들어 "변경할 좌표를 입력해주세요" 의 경우는 "LocationChangeAlert_Message" 과 같은 방식으로 네이밍을 해볼 거 같아요.

또한, 띄어쓰기보다는 언더바(_) 나 온점(.) 을 사용해 단어 사이를 구분해주는 것이 띄어쓰기 실수도 방지하고 단어 사이를 좀 더 명확히 표시해줄 수 있는 방법이라고 생각됩니다! 😊 
혹시 이 부분에 대해 어떻게 생각하실까요??

ictechgy replied:
비비가 말씀해주신게 맞는 것 같아요..! 사실 저희도 한글로 쓰면서 사용하는데에 굉장히 헷갈렸었거든요!
말씀해주신대로 변경해보겠습니다!!

YebinKim reviewed (changes requested):
코든, 샤피로 안녕하세요~
날씨정보 프로젝트(후) Step2 진행까지 고생 많으셨습니다! 👏👏
이전 스텝의 리뷰 반영도 고마워요 ☺️

로컬라이징도 `String` 의 익스텐션을 활용한 부분 정말 멋진데요! 🤩
라인별 코멘트는 구현 내용 중심으로 확인해보기로 하고, 우선 질문부터 답변 드려볼게요~



---

#### 1. API 요청 시 `CommonWeatherAPIParameter` 타입 내의 `LanguageType` 프로퍼티 사용 여부

response의 내용이 바뀌지 않은 것은 의문이네요! 🤔
저도 이 경우라면 현재 코든, 샤피로와 구현하신 방법처럼 **로컬에서 지역화**를 해주는 방법으로 작성을 했을 거 같아요~



---

그리고 현재 앱에 처음 진입했을 때 사용자의 주소를 표시하지 못 하는 문제가 있는 거 같아요!
이 부분 혹시 오류 동작인지 확인 한 번 부탁드려요 😄

<img width="320" alt="Step3" src="https://user-images.githubusercontent.com/38066039/138131321-7d5d5613-36fe-46fb-83cc-4e3d892d0a2d.PNG">


이번 스텝에서의 핵심 경험은 아래 한 가지였는데요.

<img width="224" alt="Step5" src="https://user-images.githubusercontent.com/38066039/138131414-eb63c0e8-c51e-47e2-9d31-89ae32ddd427.png">

사용자의 선호 언어에 따라 **표시 언어 및 온도 표기가 달라지는 UI 적용**과 **String 의 익스텐션을 활용한 로컬라이징 적용**에 대해서는 경험이 잘 되었다고 생각하지만,
로컬라이징 키를 사용하는 부분에 있어 현업의 스타일을 좀 더 적용해보면 좋을 거 같아 Request Changes 로 요청 드렸어요 🙏

혹시 설명이 필요한 코멘트가 있다면 언제든 DM 또는 깃헙 코멘트를 통해 알려주세요~ 😆🙌

ictechgy committed:
fix: 실 기기에서 앱 최초 구동 시 주소 정보가 올바르게 띄워지지 않던 문제 해결

- 상세 주소에 해당하는 throughfare가 없어도 주소정보가 띄워지도록 변경

shapiro711 committed:
refactor: String extension localized 메서드 수정

- comment 파라미터 추가 및 기본값 지정을 통해 좀 더 확장성있는 메서드로 변경

ictechgy committed:
style: Localization 및 API key에 대한 파일들을 폴더로 묶어서 정리

shapiro711 committed:
feat: 선호 언어 따라 지역에 맞는 온도 표기 구현

- 컨벤션 수정

ictechgy committed:
refactor: Localizable 키값들 표기방법 변경

- 직관적으로 역할을 이해할 수 있도록 Localizable.strings의 키값들을 변경

shapiro711 committed:
fix: 정확한 주소가 나오지 않던 문제 수정

- 현재 추적하고 있는지 여부를 판단하기 위해  WeatherLocationMager 타입 생성
- 수동으로 위치를 받아오는 작업시엔 위치 추적기능을 껐다가 수동으로 위치 받아오는 작업이 끝났을때 다시 켜주는 로직 구현

ictechgy committed:
fix: 셀에 시간이 잘못 표기되던 문제 수정

- 앱 최초 실행 시 requestLocation()하는 시점을 권한 허용 이후 시점으로 변경
  - requestLocation()을 했는데 사용자 권한이 허가가 계속 안되어있으면 locationManager(_:didFailWithError:)가 발생

ictechgy committed:
docs: README에 추가적으로 진행한 STEP 내용 추가

shapiro711 committed:
refactor: NetworkManager requestImage 메서드 파라미터 수정

YebinKim commented on WeatherForecast/Secret/Secret.xcconfig:
```diff
@@ -0,0 +1 @@
+API_KEY = 9a581525588c9d6997e599f7d807549b
```
API Key 가 레포에 올라왔네요 👀
의도적으로 업로드 하신 걸까요?

ictechgy replied:
헉.... ... Secret 폴더 하위로 넣으면서 `.gitignore` 경로설정과 달라지는 바람에 올라갔네요.. 이런.. 😢 신경쓰지 못했습니다. ㅠㅠ

YebinKim commented on WeatherForecast/WeatherForecast/Model/Manager/WeatherLocationManager.swift:
```diff
@@ -0,0 +1,23 @@
+import Foundation
+import CoreLocation
+
+class WeatherLocationManager: CLLocationManager {
```
만약 상속이 없을 타입이라면 final 선언하기! 잊지 맙시다 👀

ictechgy replied:
앗...또 까먹었네요..  넵....!!! 😂

YebinKim commented on WeatherForecast/WeatherForecast/Model/Manager/WeatherLocationManager.swift:
```diff
@@ -0,0 +1,23 @@
+import CoreLocation
+
+class WeatherLocationManager: CLLocationManager {
+    private(set) var isSignificantMonitoringEnabled = false
```
외부에서 변경이 불가능하도록 플래그의 접근권한을 private(set)으로 설정하셨군요! 😆

YebinKim commented on WeatherForecast/WeatherForecast/Model/Manager/AddressManager.swift:
```diff
@@ -11,31 +11,38 @@ import CoreLocation
     
     var errorDescription: String? {
         switch self {
         case .failedToGetAddress:
             return "주소값을 얻어오는데에 실패하였습니다."
         case .invalidAddress:
             return "유효한 주소가 존재하지 않습니다."
+        case .notExistPrefferedLanguage:
+            return "선호 언어가 존재하지 않습니다."
```
에러 메세지도 로컬라이징 처리를 할 수 있을 것으로 보여요~

ictechgy replied:
헉 ... 생각치도 못하고 있었네요... 😅

YebinKim commented on WeatherForecast/WeatherForecast/View/MainWeatherTableViewCell.swift:
```diff
@@ -67,20 +68,15 @@ final class MainWeatherTableViewCell: UITableViewCell {
-        dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
-        
-        if let forecastedDate = data.timeOfDataForecasted, let date = dateFormatter.date(from: forecastedDate), let preferredLanguage = Locale.preferredLanguages.first {
+        if let forecastedDate = data.timeOfDataCalculation, let preferredLanguage = Locale.preferredLanguages.first {
```
`forecastedDate` 는 변경 전과 변수명이 달라지지 않은 것으로 보이네요 🤔
의도된 네이밍이 맞을까요?

ictechgy replied:
저는 `calculatedDate`라는 네이밍을 제안했었는데 Shapiro는 이 네이밍이 맞을 것 같다는 쪽의 의견을 내주셨었어요!

YebinKim replied:
의도한 네이밍이 맞았군요!

YebinKim commented on WeatherForecast/WeatherForecast/Controller/MainWeatherViewController.swift:
```diff
@@ -83,15 +91,17 @@ extension MainWeatherViewController: CLLocationManagerDelegate {
     func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
-        print(error)
+        debugPrint(error)
+        tableView.refreshControl?.endRefreshing()
```
UI 관련 작업은 메인 스레드에서의 실행이 보장되어야 하지 않을까요? 🤔

ictechgy replied:
앗.. "이 메서드는 당연히 메인 쓰레드에서 실행되지 않을까" 하는 생각을 했었는데요.. 안정성을 위해 특별한 이유가 없는 한 메인쓰레드 실행 보장을 해주는 것이 좋을 것 같네요..!

YebinKim commented on README.md:
```diff
@@ -1,8 +1,22 @@
 
-- 프로젝트 기간: 2021년 9월 27일 ~ ing
+- 프로젝트 기간: 2021년 9월 27일 ~ 10월 22일
 - 프로젝트 진행자: Coden, Shapiro
```
프로젝트 진행자에 깃헙 아이디를 함께 적어줘도 좋을 거 같아요 😆

ictechgy replied:
핫 넵 알겠습니다.. ㅎㅎ

YebinKim commented on README.md:
```diff
@@ -300,3 +314,381 @@
+
+- 최초 실행 시 좌표값과 주소값에 대한 정보는 아래와 같았다.
+
+<img width="900" src="https://user-images.githubusercontent.com/39452092/138454159-b518e35f-b824-4dd1-a747-a14c7a93a979.png" />
```
이미지가 너무 커서 내용 파악에 어려움이 약간 있네요!
img width 값을 줄여주면 동일한 비율로 이미지의 크기를 줄여줄 수 있어요~

ictechgy replied:
저희는 이정도 이미지 크기가 괜찮은 것 같다고 생각했었는데 아니었나보군요..! 😂

YebinKim replied:
앗 요거 제가 라인을 잘못 잡았네요..!
모바일 화면의 캡쳐 이미지가 너무 크다고 생각했어요~ 코드 캡쳐 이미지 사이즈는 적당한 거 같습니다!

YebinKim commented on README.md:
```diff
@@ -300,3 +314,381 @@
+
+## 💫 TroubleShooting
+
+1. `MainWeatherViewController`에서 `MainWeatherTableViewCell`에 데이터를 설정해주는 방식에 대하여
```
질문의 개수가 많은데 질문과 내용의 위계가 같아서 가독성이 조금 떨어지는 거 같아요!
질문은 볼드처리 또는 ### 로 위계 수준을 올려주면 어떨까요?

ictechgy replied:
😂🥲 알겠습니다!!

YebinKim commented on README.md:
```diff
@@ -300,3 +314,381 @@
+   * 우리는 CollectionView를 직접적으로 사용하는 부분이 없다.
+   * 이에 대해 alert 및 alert의 textFields들에 대한 `translateAutoResizingMaskIntoConstraint` 설정을 true로 했다가 false로도 해봤지만 소용이 없었다. (오히려 alert가 기괴하게 변하였음)
+
+**2021-10-22 18:08:34.439513+0900 WeatherForecast[3447:1559084] [LayoutConstraints] Changing the translatesAutoresizingMaskIntoConstraints property of a UICollectionViewCell that is managed by a UICollectionView is not supported, and will result in incorrect self-sizing. View: <_UIAlertControllerTextFieldViewCollectionCell: 0x104a60460; frame = (0 0; 270 24); gestureRecognizers = <NSArray: 0x2814 ...
+
+**2021-10-22 18:08:34.440665+0900 WeatherForecast[3447:1559084] Writing analzed variants.**
+
+**2021-10-22 18:08:34.442111+0900 WeatherForecast[3447:1559084] [LayoutConstraints] Changing the translatesAutoresizingMaskIntoConstraints property of a UICollectionViewCell that is managed by a UICollectionView is not supported, and will result in incorrect self-sizing. View: <_UIAlertControllerTextFieldViewCollectionCell: 0x104b341d0; frame = (0 24; 270 24); gestureRecognizers = <NSArray: 0x281 ...
+
+**2021-10-22 18:08:34.443500+0900 WeatherForecast[3447:1559084] [UICollectionViewRecursion] cv == 0x106011e00 Disabling recursion trigger logging**
+
+**2021-10-22 18:08:34.447389+0900 WeatherForecast[3447:1559084] Writing analzed variants.**
+
+**2021-10-22 18:08:34.504177+0900 WeatherForecast[3447:1559084] Can't find keyplane that supports type 8 for keyboard iPhone-PortraitChoco-DecimalPad; using 27100_PortraitChoco_iPhone-Simple-Pad_Default**
```
오류 내용은 접어두거나 코드 블럭으로 감싸서 내용과 다르게 표시해주는게 이해하기 좋을 거 같아요 😊

ictechgy replied:
알겠습니다..!! ㅎㅎ.. 🥲

YebinKim reviewed (approved):
샤피로, 코든 안녕하세요~
날씨정보 프로젝트(후)  Step2 수정 진행까지 고생 많으셨습니다!
지난 목요일 지역 정보 상세를 표시하는 것에 대해 얘기를 나누다가 Refresh Control 이 무한 로딩되는 문제와 날씨 예보 시간과 아이콘이 매칭되지 않았던 문제를 갑자기 발견하게 되었었는데요.
이 부분까지 포함해서 더 완성도 있게 스텝을 진행해주신 점 정말 멋집니다! 🤩👍

프로젝트 리드미는 각 스텝별로 구현하며 발생했던 문제와 고민이 아주 잘 녹아있었는데요! 👍
**프로젝트의 기능 설명과 기술 스택을 문서 초반에 추가**함으로써 프로젝트를 처음 보는 사람도 이해할 수 있도록 작성하고,
**프로젝트 목차 또한 추가**해서 스텝별 진행사항을 링크로 빠르게 찾아갈 수 있게 해두면 더 읽기 편할 거 같다는 생각이 들었어요~

각 스텝별 트러블 슈팅은 현재와 같은 문장식 설명 보다는 문제 상황과 해결 방법을 구분해서 개괄식으로 정리하면 좀 더 깔끔하게 볼 수 있을 거 같다는 생각이 들었는데요!
문서적인 부분은 다른 리드미 문서도 참고해보고 또 이번에 작성했던 리드미를 2달 뒤 다시 읽어보면서 이해하기 쉬운 문서 스타일을 찾아가보시면 좋을 거 같습니다 😊

내용은 PR 내용과 크게 다르지 않아서 따로 피드백 드릴건 없을 거 같아요! ㅎㅎ

지난 4주간 고생 많으셨습니다 🥳🎉
